docs/strings.md

```diff
@@ -0,0 +1,464 @@
+
+	/**
+	 * Replace the specified range in a string with another string.
+	 */
+	func update_range[String Str](self: (mut | iso) & String,
+	                              range: Range,
+	                              newString: Str & imm);
+}
+```
+
+For discussion, the above proposal is somewhat simplified:
+The `update` operation should take a `StringView`, a non-owning view of a substring.
+The `String` interface should also include an operation for providing a `StringView` on a specific range.
+The Verona region system makes tracking ownership of string views relatively simple:
+
+ - By default, all mutable strings are `iso`.
+ - Mutable string views are `mut`, in the same region as their owning string.
+ - Immutable strings are globally immutable (`imm`).
+ - Immutable string views are immutable.
+
+This makes it easy to ensure that the lifetime of a string view is bound to the lifetime of the underlying string.
+Once stack types are better defined, it should also be possible for string views to be stack types associated with the region of the string.
+
+The minimal standard library should also include:
+
+ - Concrete instantiations of the `String` interface, storing ASCII, UTF-8, and UTF-16.
+ - Operations for finding substrings and extracting substrings as string views.
+ - Operations for constructing the most size or speed-efficient concrete string type from a string literal or string view.
+ - Operations for computing efficient hashes on strings.
+
+Requirements on the language
+----------------------------
+
+Supporting a rich Verona string library imposes a few requirements on the language.
+Most of these are already in progress but are captured here for completeness.
+
+# Literal syntax
+
+String literals in the source code should be an opaque immutable type that implements an interface that allows bytes to be iterated.
+We then need lightweight syntax for passing an instance of this type to the constructor for a concrete string type.
+
+# Generics
+
+Most string functions will be generics that take a `String` or `StringView` interface but then match on concrete instantiations of that interface for fast paths.
+The concrete type of a string will often be known to the compiler but not to the static type system.
+As such, we want simple syntax for writing generics and a language-level guarantee that the compiler is free to reify and call more-specific versions if the type is known statically.
+For example, consider a trivial search function:
+
+```verona
+func find_character(str : String & readonly, c : Unichar) : U64 | NotFound
+{
+	match (str)
+	{
+		AsciiString => (if (c > 127) return NotFound) else ...
+		...
+	}
+}
+```
+
+This should be equivalent to a generic where the type used for `str` is a generic paramter.
+If the compiler knows that the argument is an `AsciiString` then it should be free to substitute a version that only includes the `AsciiString` version of the `match`.
```

Comment:
This is what I think @sylvanc was referring to earlier. If we convert that to a template parameter, restricted to implement the `String` interface, it becomes easier to match the type at compile time:
```
func(a: T & readonly, c: Unichar) : U64 | NotFound
    where T: String
{
    ...
}
```

Reply:
Yes, exactly.  This is a write-up of what @syvanc and I discussed.